### PR TITLE
feat(pipx support): Add support for pipx

### LINKED ISSUE #48: Add support for pipx?
Hello, is it possible to add support for pipx installation utility? It is the recommended method to install python executables that are not available in the default repositories or in AUR. 

Thanks a lot!

### PR DESCRIPTION
**Pull request by InnocentZero**
This is a series of commits intended to add support for pipx in pacdef. WIP. Closes #48

### CODE REVIEW

File: crates/pacdef_core/src/backend/actual/python.rs [hunk 1]
```diff
@@ -57,9 +67,18 @@ fn run_pip_command(args: &[&str]) -> Result<Value> {
 impl Python {
     pub(crate) fn new() -> Self {
         Self {
+            binary: BINARY.to_string(),
             packages: HashSet::new(),
         }
     }
+
+    fn get_switches_runtime(&self) -> Switches {
+        if self.get_binary().eq("pip") {
```
Comment (steven-omaha) [inline crates/pacdef_core/src/backend/actual/python.rs:76]: please `match` the String against `pip` and `pipx`, and `panic!()` if the user supplies anything else.

Comment (InnocentZero) [reply crates/pacdef_core/src/backend/actual/python.rs]: > please `match` the String against `pip` and `pipx`, and `panic!()` if the user supplies anything else.

This has been implemented in the recent commit. Please have a look.

**Comment by InnocentZero** [comment]
@steven-omaha I think I need your help on this one.

So the way pipx works is that in its JSON, it first lists the primary packages installed and then their dependencies. I'm trying to iterate through the dependencies but there is a type mismatch for some reason?

Here is a JSON generated through pipx

```json
{
    "pipx_spec_version": "0.1",
    "venvs": {
        "mdformat-myst": {
            "metadata": {
                "injected_packages": {},
                "main_package": {
                    "app_paths": [],
                    "app_paths_of_dependencies": {
                        "markdown-it-py": [
                            {
                                "__Path__": "/home/innocentzero/.local/share/pipx/venvs/mdformat-myst/bin/markdown-it",
                                "__type__": "Path"
                            }
                        ],
                        "mdformat": [
                            {
                                "__Path__": "/home/innocentzero/.local/share/pipx/venvs/mdformat-myst/bin/mdformat",
                                "__type__": "Path"
                            }
                        ]
                    },
                    "apps": [],
                    "apps_of_dependencies": [
                        "mdformat",
                        "markdown-it"
                    ],
                    "include_apps": true,
                    "include_dependencies": true,
                    "man_pages": [],
                    "man_pages_of_dependencies": [],
                    "man_paths": [],
                    "man_paths_of_dependencies": {},
                    "package": "mdformat-myst",
                    "package_or_url": "mdformat-myst",
                    "package_version": "0.1.5",
                    "pip_args": [],
                    "suffix": ""
                },
                "pipx_metadata_version": "0.3",
                "python_version": "Python 3.11.6",
                "venv_args": []
            }
        },
    }
}
``` 

`app_paths_of_dependencies` is where all the dependencies are listed as keys.

Please have a look at the function commented at the bottom of the file. Thanks.

**Comment by steven-omaha** [comment]
@InnocentZero I don't think that code is needed, unless my understanding is wrong. AFAIU pipx installs every binary package in a separate venv with all its dependencies. That guarantees that all dependencies are always fulfilled, and that nothing else is installed in the corresponding venv, unless you do something like `source $HOME/.local/share/pipx/venv/somepackage/bin/activate; pip install somethingelse`. 

I have not used pipx before. If manually adding something to a venv is a valid use case, then we can and should look at that code, otherwise `get_all_installed_packages` and `get_explicitly_installed_packages` should produce the same output for pipx, which means by definition that pipx does not have unmanaged packages.

**Comment by steven-omaha** [comment]
I have encountered something similar when I implemented this for `cargo install` in `rust.rs`. Feel free to take a look at the code there.

**Comment by InnocentZero** [comment]
@steven-omaha it _kinda_ fully supports `pipx` now. 

Now there are some packages like `mdformat-myst`, that themselves do not contain executables but instead act as a library for their dependencies, which is `mdformat` in this case. 

The rationale behind this is `mdformat` is core markdown spec compliant, while `mdformat-myst` is an additional layer of features that are not a part of the core. So it only acts as a library to `mdformat`.

To install such packages, an additional `--include-deps` flag needs to be passed to it. However, it cannot be added willy-nilly because in cases of apps like `yewtube` (or any other for that matter), it adds a whole lot of unnecessary executables. So it would be great if pacdef can give a warning when using `pipx` that the flag may be necessary in some circumstances.

**Comment by InnocentZero** [comment]
> @InnocentZero I don't think that code is needed, unless my understanding is wrong. AFAIU pipx installs every binary package in a separate venv with all its dependencies. That guarantees that all dependencies are always fulfilled, and that nothing else is installed in the corresponding venv, unless you do something like `source $HOME/.local/share/pipx/venv/somepackage/bin/activate; pip install somethingelse`.
> 
> I have not used pipx before. If manually adding something to a venv is a valid use case, then we can and should look at that code, otherwise `get_all_installed_packages` and `get_explicitly_installed_packages` should produce the same output for pipx, which means by definition that pipx does not have unmanaged packages.

Ah I see. Then I think pipx already does a decent job at it. That functionality is unneeded.

File: crates/pacdef_core/src/backend/actual/python.rs [hunk 2]
```diff
@@ -57,9 +70,25 @@ fn run_pip_command(args: &[&str]) -> Result<Value> {
...
+            binary: BINARY.to_string(),
             packages: HashSet::new(),
         }
     }
+
+    fn get_switches_runtime(&self) -> Switches {
+        match self.get_binary() {
+            "pip" => &["list", "--format", "json", "--not-required", "--user"],
+            "pipx" => &["list", "--json"],
+            _ => panic!("Cannot use {} for package management in python. Please use a valid package manager like pip or pipx", self.get_binary()),
```
Comment (steven-omaha) [inline crates/pacdef_core/src/backend/actual/python.rs:82]: Please extract the message as a constant as far as possible. This message appears four times in this module.

File: crates/pacdef_core/src/backend/actual/python.rs [hunk 3]
```diff
@@ -29,25 +30,37 @@ const SUPPORTS_AS_DEPENDENCY: bool = false;
...
+    fn get_binary(&self) -> Text {
+        let r#box = self.binary.clone().into_boxed_str();
+        Box::leak(r#box)
+    }
 
-        extract_pacdef_packages(output)
+    fn get_all_installed_packages(&self) -> Result<HashSet<Package>> {
+        let mut cmd = Command::new(self.get_binary());
+        let output = run_pip_command(&mut cmd, self.get_switches_runtime())?;
+        match self.get_binary(){
```
Comment (steven-omaha) [inline crates/pacdef_core/src/backend/actual/python.rs:41]: Lines 41 to 45 are identical with 51 to 55. Please extract function or method.

**Review by steven-omaha — COMMENTED** [review]
Very good work overall. I have added two comments where the code could be improved. Aside from this the code is ready to be merged.

File: crates/pacdef_core/src/backend/actual/python.rs (hunk 2 above)
Comment (InnocentZero) [reply crates/pacdef_core/src/backend/actual/python.rs]: Done!

**Comment by steven-omaha** [comment]
Well done. Thank you so much!

**Comment by steven-omaha** [comment]
@InnocentZero I'd appreciated some comments on how easy or hard it was to add your feature to pacdef. Is there anything in the program that you think could be improved?

**Pull request merged**